### PR TITLE
Move Build onto the injection-gen based controller infrastructure.

### PR DESCRIPTION
**Pull request by mattmoor**
I've broken this into finer grained commits as a sort guide for migration.  I can try to take a crack at doing something along these lines for tekton as well, but the goal of the broader refactoring was to effectively eliminate boilerplate outside of `pkg/reconciler/foo` for writing controllers, and keep our various controllers from drifting apart as we add features (e.g. this gives Build various metrics integrations including stackdriver that I believe it lacked previously).

This also enables us to pretty trivially consolidate controllers, or split them apart by simply changing what goes into the main binary.

/hold

cc @bobcatfish @vdemeester @n3wscott

### CODE REVIEW

File: config/config-observability.yaml [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline config/config-observability.yaml]:
```suggestion
# Copyright 2019 The Knative Authors
```

File: pkg/client/injection/client/client.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/client/client.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/client/fake/fake.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/client/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/factory/buildfactory.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/factory/buildfactory.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/factory/fake/fake.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/factory/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/build/build.go [hunk 6]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/build/build.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/build/fake/fake.go [hunk 7]
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/build/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/buildtemplate/buildtemplate.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/buildtemplate/buildtemplate.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/buildtemplate/fake/fake.go [hunk 9]
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/buildtemplate/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/clusterbuildtemplate.go [hunk 10]
```diff
@@ -0,0 +1,52 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/clusterbuildtemplate.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/fake/fake.go [hunk 11]
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/reconciler/build/controller.go [hunk 12]
```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/reconciler/build/controller.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/reconciler/buildtemplate/controller.go [hunk 13]
```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/reconciler/buildtemplate/controller.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/reconciler/clusterbuildtemplate/controller.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/reconciler/clusterbuildtemplate/controller.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: config/config-observability.yaml (hunk 1 above)
Comment (mattmoor-sockpuppet) [inline config/config-observability.yaml]:
```suggestion
# Copyright 2019 The Knative Authors
```

File: pkg/client/injection/client/client.go (hunk 2 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/client/client.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/client/fake/fake.go (hunk 3 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/client/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/factory/buildfactory.go (hunk 4 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/factory/buildfactory.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/factory/fake/fake.go (hunk 5 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/factory/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/build/build.go (hunk 6 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/build/build.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/build/fake/fake.go (hunk 7 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/build/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/buildtemplate/buildtemplate.go (hunk 8 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/buildtemplate/buildtemplate.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/buildtemplate/fake/fake.go (hunk 9 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/buildtemplate/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/clusterbuildtemplate.go (hunk 10 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/clusterbuildtemplate.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/fake/fake.go (hunk 11 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/injection/informers/build/v1alpha1/clusterbuildtemplate/fake/fake.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/reconciler/build/controller.go (hunk 12 above)
Comment (mattmoor-sockpuppet) [inline pkg/reconciler/build/controller.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/reconciler/buildtemplate/controller.go (hunk 13 above)
Comment (mattmoor-sockpuppet) [inline pkg/reconciler/buildtemplate/controller.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/reconciler/clusterbuildtemplate/controller.go (hunk 14 above)
Comment (mattmoor-sockpuppet) [inline pkg/reconciler/clusterbuildtemplate/controller.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: config/config-observability.yaml (hunk 1 above)
Comment (mattmoor-sockpuppet) [inline config/config-observability.yaml]:
```suggestion
# Copyright 2019 The Knative Authors
```

**Comment by vdemeester** [comment]
@mattmoor nice ! I'll look into that for `tekton` :angel:

File: pkg/reconciler/build/build_test.go [hunk 15]
```diff
@@ -23,22 +23,28 @@ import (
...
 	corev1 "k8s.io/api/core/v1"
 	kuberrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
-	kubeinformers "k8s.io/client-go/informers"
-	k8sfake "k8s.io/client-go/kubernetes/fake"
 	clientgotesting "k8s.io/client-go/testing"
 	"k8s.io/client-go/tools/cache"
+
+	. "github.com/knative/pkg/reconciler/testing"
```
Comment (ImJasonH) [inline pkg/reconciler/build/build_test.go]: Ew, `.` imports. Can we avoid this and explicitly name things imported from `testing` ?

Comment (mattmoor) [reply pkg/reconciler/build/build_test.go]: Done.

**Comment by mattmoor** [comment]
@vdemeester let me of @n3wscott know if you run into any trouble :)

**Pull request merged**